File: 2903-insert-greatest-common-divisors-in-linked-list/insert-greatest-common-divisors-in-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func findGCD(a int, b int) int {
    for a != b {
        if a > b {
            a -= b
        } else if b > a {
            b -= a
        }
    }
    return a
 }
func insertGreatestCommonDivisors(head *ListNode) *ListNode {
    current := head

    for current != nil && current.Next != nil {
        gcd := findGCD(current.Val, current.Next.Val)
        newNode := new(ListNode)
        newNode.Val = gcd
        newNode.Next = current.Next
        current.Next = newNode

        current = newNode.Next
    }

    return head
}

File: 2903-insert-greatest-common-divisors-in-linked-list/insert-greatest-common-divisors-in-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    private int findGCD(int a, int b) {
        while (a != b) {
            if (a > b) a -= b;
            else if (b > a) b -= a;
        }

        return a;
    }

    public ListNode insertGreatestCommonDivisors(ListNode head) {
        for (ListNode cur = head; cur != null && cur.next != null; cur = cur.next.next) {
            int gcd = findGCD(cur.val, cur.next.val);
            ListNode newNode = new ListNode(gcd, cur.next);
            cur.next = newNode;
        }

        return head;
    }
}